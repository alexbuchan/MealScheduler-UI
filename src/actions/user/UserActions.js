require('es6-promise').polyfill();
const request = require('axios');

import ActionDispatch from './ActionDispatch';
import FlashMessageActions from '../FlashMessageActions/FlashMessageActions';
import ActionsHelper from '../ActionsHelper';
import JWT from '../../lib/JWT/JWT';

class UserActions {
  registerUser = async user => {
    const _endpoint = 'http://localhost:3000/signup';

    let error, response;
    [error, response] = await ActionsHelper.asyncHelper(
      request.post(_endpoint, user)
    );
    
    if (error) {
      FlashMessageActions.dispatchErrorMessage(error.response);
    } else {
      this.handleSignupLoginResponse(response, ActionDispatch.dispatchRegisterUser);
    }
  }

  loginUser = async user => {
    const _endpoint = 'http://localhost:3000/login';

    let error, response;
    [error, response] = await ActionsHelper.asyncHelper(
      request.post(_endpoint, user)
    );
    
    if (error) {
      FlashMessageActions.dispatchErrorMessage(error.response);
    } else {
      this.handleSignupLoginResponse(response, ActionDispatch.dispatchLoginUser);
    }
  }

  logoutUser = () => {
    ActionsHelper.removeCookie();
    ActionDispatch.dispatchLogoutUser();
  }

  retrieveUserDataOnRefresh = () => {
    const cookie = ActionsHelper.getCookie('user');
    if (cookie) {
      const data = JWT.decodeJWTToken(cookie);
      ActionDispatch.dispatchUserDataOnRefresh(data);
    }
  }

  handleSignupLoginResponse = (response, dispatchFunction) => {
    if (response.status === 200) {
      const data = JWT.decodeJWTToken(response.data.token);
      ActionsHelper.setCookie('user', response.data.token);

      dispatchFunction(data);
    }
  }
}

export default new UserActions();
