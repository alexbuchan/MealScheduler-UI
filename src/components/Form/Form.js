import React from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import Validator from '../../lib/Validator';

const propTypes = {
  submitButtonLabel: PropTypes.string,  // Provide a label name for the submit button
  validate: PropTypes.bool,             // Should the form be validated
  fields: PropTypes.object.isRequired,  // Provide the fields for validation
  onSubmit: PropTypes.func.isRequired,  // Provide the onSubmit function to handle form submittion,
  redirectTo: PropTypes.string,         // If redirect enabled, pass route to be redirected to
  shouldRedirect: PropTypes.func,       // If redirect enabled, pass conditions to Form for redirect
};

class Form extends React.Component {
  constructor(props) {
    super(props);

    const validationFields = this.props.children.map(child => child.props.name);
    this.validator = new Validator(...validationFields);

    this.state = {
      redirect: false,
      disableSubmitButton: false,
      validation: this.validator.valid()
    }

    this.submitted = false;
  }

  componentDidUpdate() {
    if (this.state.disableSubmitButton) {
      this.turnOffSubmitButtonDisable = setTimeout(() => {
        this.setState(() => ({ disableSubmitButton: false }))
      }, 1000);
    }

    this.redirect();
  }

  handleOnSubmit = (ev) => {
    ev.preventDefault();

    if (this.props.validate) {
      const validation = this.validator.validate(this.props.fields);
      this.setState({ validation });
      this.submitted = true;
  
      if (validation.isValid) {
        this.props.onSubmit(this.props.fields);
        this.disableSubmitButtonOnSubmit();
      }
    } else {
      this.props.onSubmit(this.props.fields);
      this.disableSubmitButtonOnSubmit();
    }
  }

  disableSubmitButtonOnSubmit = () => {
    this.setState({
      disableSubmitButton: true
    });
  }

  redirect = () => {
    if (this.props.shouldRedirect()) {
      this.setState({ redirect: true });
    }
  }

  render() {
    if (this.state.redirect) return <Redirect to={ this.props.redirectTo } />;

    let validation = this.submitted ? this.validator.validate(this.props.fields) : this.state.validation;
    const children = React.Children.toArray(this.props.children);
    const clonedChildren = children.map(child => React.cloneElement(child, { validationField: validation[child.props.name] }));

    return (
      <div className='my-form'>
        { clonedChildren }
        <button
          className="form-submit-button"
          disabled={ this.state.disableSubmitButton }
          onClick={ this.handleOnSubmit }
        >
          { this.props.submitButtonLabel }
        </button>
      </div>
    );
  }
}

Form.defaultProps = {
  submitButtonLabel: 'Submit',
  validate: false
}

Form.propTypes = propTypes;
export default Form;