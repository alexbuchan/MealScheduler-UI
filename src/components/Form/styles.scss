.my-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
}

.form-submit-button {
  width: 178px;
  height: 40px;
  margin-top: 30px;
  background-image: linear-gradient(0deg, #4db1f3, #b0d8f3);
  border: 1px solid #4575ad;
  border-radius: 4px;
}

.form-link {
  font-size: small;
  margin-top: 10px;
}