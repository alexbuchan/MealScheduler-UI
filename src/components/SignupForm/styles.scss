.help-block {
  color: #fb2a2a;
  font-size: 12px;
  text-align: center;
}

.input-field-error {
  border: 1px solid #fb2a2a;
}

.input-field-error:focus {
  outline: 1px solid rgba($color: #fb2a2a, $alpha: 0.2);
}