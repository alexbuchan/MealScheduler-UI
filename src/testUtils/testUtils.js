import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}