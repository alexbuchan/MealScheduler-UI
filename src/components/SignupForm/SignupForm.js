import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import UserActions from '../../actions/user/UserActions';
import Form from '../Form/Form';
import TextField from '../formComponents/TextField/TextField';

const propTypes = {
  user: PropTypes.object   // User prop makes sure once registration has taken place that form can redirect.
}


class SignupForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: '',
      email: '',
      password: ''
    };
  }

  handleOnChange = (ev) => {
    ev.preventDefault();

    this.setState({
      [ev.target.name]: ev.target.value
    });
  }

  handleOnSubmit = () => {
    UserActions.registerUser(this.state);
  }

  formRedirect = () => {
    if (this.props.user) return true;
    else return false;
  }

  render() {
    return (
      <div className='user-signup'>
        <div className='signup-form'>
          <Form
            submitButtonLabel='Register!'
            validate={ true }
            fields={ this.state }
            areRequired={ ['username', 'email', 'password'] }
            onSubmit={ this.handleOnSubmit }
            redirect={ true }
            redirectTo='/schedule'
            shouldRedirect={ this.formRedirect() }
          >
            <TextField
              label='Username'
              name='username'
              value={ this.state.username }
              onChange={ this.handleOnChange }
            />

            <TextField
              label='Email'
              name='email'
              value={ this.state.email }
              onChange={ this.handleOnChange }
            />

            <TextField
              label='Password'
              name='password'
              type='password'
              value={ this.state.password }
              onChange={ this.handleOnChange }
            />
          </Form>
          <p className="form-link">Already have an account? <Link className="" to="/login">Log in</Link></p>
        </div>
      </div>
    );
  }
}

SignupForm.propTypes = propTypes;
export default SignupForm;
