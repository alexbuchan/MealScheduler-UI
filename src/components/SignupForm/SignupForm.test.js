import SignupForm from './SignupForm';
import Form from '../Form/Form';
import UserActions from '../../actions/user/UserActions';
import { componentSetup } from '../../testUtils/testUtils';

jest.mock('../../actions/user/UserActions');

describe('SignupForm', () => {
  let instance;

  beforeEach(() => {
    const props = {
      user: undefined
    }

    instance = componentSetup(SignupForm, props);
  });

  it('renders correctly', () => {
    expect(instance).toMatchSnapshot();
  });

  describe('#handleOnChange', () => {
    it('sets the state to input value each time user inputs a value', () => {
      const preventDefault = jest.fn();
      const emailTextField = instance.find({ name: 'email' });

      emailTextField.simulate("change", { target: { name: 'email', value: 'email' }, preventDefault: preventDefault });

      expect(instance.state().email).toBe('email');
    });
  });

  describe('#formRedirect', () => {
    describe('when user is NOT registered', () => {
      let instance;

      beforeEach(() => {
        const props = {
          user: undefined
        }

        instance = componentSetup(SignupForm, props);
      });

      it('returns false', () => {
        expect(instance.instance().formRedirect()).toBe(false);
      });
    });

    describe('when user is registered', () => {
      let instance;

      beforeEach(() => {
        const props = {
          user: { username: 'username', email: 'user@email.com' }
        }

        instance = componentSetup(SignupForm, props);
      });

      it('returns true', () => {
        expect(instance.instance().formRedirect()).toBe(true);
      });
    });
  });

  describe('#handleOnSubmit', () => {
    it('calls the user register action', () => {
      const preventDefault = jest.fn();
      const form = instance.find(Form);

      instance.instance().setState({
        username: 'username',
        email: 'user@email.com',
        password: 'password'
      });

      form.simulate("submit", { preventDefault: preventDefault });

      expect(UserActions.registerUser).toHaveBeenCalledTimes(1);
      expect(UserActions.registerUser).toHaveBeenCalledWith({
        username: 'username',
        email: 'user@email.com',
        password: 'password'
      });
    });
  });
});