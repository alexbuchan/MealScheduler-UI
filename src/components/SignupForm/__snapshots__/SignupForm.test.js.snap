// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm renders correctly 1`] = `
<div
  className="user-signup"
>
  <div
    className="signup-form"
  >
    <Form
      areRequired={
        Array [
          "email",
          "password",
        ]
      }
      fields={
        Object {
          "email": "",
          "password": "",
          "username": "",
        }
      }
      onSubmit={[Function]}
      redirect={true}
      redirectTo="/contacts"
      shouldRedirect={false}
      submitButtonLabel="Register!"
      validate={true}
    >
      <TextField
        label="Username"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
      <TextField
        label="Email"
        name="email"
        onChange={[Function]}
        type="text"
        value=""
      />
      <TextField
        label="Password"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </Form>
    <p
      className="form-link"
    >
      Already have an account? 
      <Link
        className=""
        to="/login"
      >
        Log in
      </Link>
    </p>
  </div>
</div>
`;
